server/migrator/Schema.java

```diff
@@ -0,0 +1,181 @@
+                        .append(String.format("relates %s", x.getLabel().name())));
+    }
+
+    private void writePlays(StringBuilder builder, ThingType thingType) {
+        thingType.getPlays().sorted(Comparator.comparing(x -> x.getLabel().scopedName()))
+                .forEach(x -> builder
+                        .append(COMMA_NEWLINE_INDENT)
+                        .append(String.format("plays %s", x.getLabel().scopedName())));
+    }
+
+    private void writeRule(StringBuilder builder, Rule rule) {
+        builder.append(String.format("%s sub rule", rule.getLabel()))
+                .append(COMMA_NEWLINE_INDENT)
+                .append("when\n")
+                .append(getPatternString(wrapConjunction(rule.getWhenPreNormalised()), 1))
+                .append(COMMA_NEWLINE_INDENT)
+                .append("then\n")
+                .append(getPatternString(wrapConjunction(rule.getThenPreNormalised()), 1))
+                .append(SEMICOLON_NEWLINE_X2);
+    }
+
+    private String getValueTypeString(AttributeType.ValueType valueType) {
+        switch (valueType) {
+            case STRING:
+                return "string";
+            case LONG:
+                return "long";
+            case DOUBLE:
+                return "double";
+            case BOOLEAN:
+                return "boolean";
+            case DATETIME:
+                return "datetime";
+            default:
+                throw GraknException.of(ILLEGAL_STATE);
+        }
+    }
+
+    private String getPatternString(Pattern pattern, int indent) {
+        if (pattern.isVariable()) {
+            return indent(indent) + pattern.asVariable().toString();
+        } else if (pattern.isConjunction()) {
+            StringBuilder builder = new StringBuilder()
+                    .append(indent(indent))
+                    .append("{\n");
+            pattern.asConjunction().patterns().forEach(p -> builder
+                    .append(getPatternString(p, indent + 1))
+                    .append(";\n"));
+            builder.append(indent(indent))
+                    .append("}");
+            return builder.toString();
+        } else if (pattern.isDisjunction()) {
+            return pattern.asDisjunction().patterns().stream()
+                    .map(p -> getPatternString(wrapConjunction(p), indent))
+                    .collect(Collectors.joining("\n" + indent(indent) + "or\n"));
+        } else if (pattern.isNegation()) {
+            return indent(indent) + "not\n" + getPatternString(wrapConjunction(pattern.asNegation().pattern()), indent);
+        } else {
+            throw GraknException.of(ILLEGAL_STATE);
+        }
+    }
+
+    private Pattern wrapConjunction(Pattern pattern) {
+        return pattern.isConjunction() ? pattern : new Conjunction<>(list(pattern));
+    }
+
+    private static String indent(int indent) {
+        return IntStream.range(0, indent * 4).mapToObj(i -> " ").collect(Collectors.joining(""));
```

Comment:
I think we can replace the last `collect` with `Collectors.joining()`?

Reply:
ok